pkg/minikube/bootstrapper/kubeadm/kubeadm.go

```diff
@@ -638,124 +583,41 @@ func (k *Bootstrapper) restartControlPlane(cfg config.ClusterConfig) error {
-		return errors.Wrap(err, "cp")
-	}
-
-	baseCmd := fmt.Sprintf("%s %s", bsutil.InvokeKubeadm(cfg.KubernetesConfig.KubernetesVersion), phase)
-	cmds := []string{
-		fmt.Sprintf("%s phase certs all --config %s", baseCmd, conf),
-		fmt.Sprintf("%s phase kubeconfig all --config %s", baseCmd, conf),
-		fmt.Sprintf("%s phase kubelet-start --config %s", baseCmd, conf),
-		fmt.Sprintf("%s phase %s all --config %s", baseCmd, controlPlane, conf),
-		fmt.Sprintf("%s phase etcd local --config %s", baseCmd, conf),
-	}
-
-	klog.Infof("reconfiguring cluster from %s", conf)
-	// Run commands one at a time so that it is easier to root cause failures.
-	for _, c := range cmds {
-		if _, err := k.c.RunCmd(exec.Command("/bin/bash", "-c", c)); err != nil {
-			klog.Errorf("%s failed - will try once more: %v", c, err)
 
-			if _, err := k.c.RunCmd(exec.Command("/bin/bash", "-c", c)); err != nil {
-				return errors.Wrap(err, "run")
-			}
-		}
+	if rr, err := k.c.RunCmd(exec.Command("sudo", "diff", "-u", conf, conf+".new")); err != nil {
+		klog.Infof("needs reconfigure: configs differ:\n%s", rr.Output())
+		return true
 	}
-
-	cr, err := cruntime.New(cruntime.Config{Type: cfg.KubernetesConfig.ContainerRuntime, Runner: k.c})
+	// cruntime.Enable() may restart kube-apiserver but does not wait for it to return back
+	apiStatusTimeout := 3000 * time.Millisecond
+	st, err := kverify.WaitForAPIServerStatus(k.c, apiStatusTimeout, hostname, port)
 	if err != nil {
-		return errors.Wrap(err, "runtime")
-	}
-
-	// We must ensure that the apiserver is healthy before proceeding
-	if err := kverify.WaitForAPIServerProcess(cr, k, cfg, k.c, time.Now(), kconst.DefaultControlPlaneTimeout); err != nil {
-		return errors.Wrap(err, "apiserver healthz")
-	}
-
-	if err := kverify.WaitForHealthyAPIServer(cr, k, cfg, k.c, client, time.Now(), hostname, port, kconst.DefaultControlPlaneTimeout); err != nil {
-		return errors.Wrap(err, "apiserver health")
-	}
-
-	// because reboots clear /etc/cni
-	if err := k.applyCNI(cfg); err != nil {
-		return errors.Wrap(err, "apply cni")
-	}
-
-	if err := kverify.WaitForSystemPods(cr, k, cfg, k.c, client, time.Now(), kconst.DefaultControlPlaneTimeout); err != nil {
-		return errors.Wrap(err, "system pods")
-	}
-
-	if err := kverify.NodePressure(client); err != nil {
-		adviseNodePressure(err, cfg.Name, cfg.Driver)
-	}
-
-	// This can fail during upgrades if the old pods have not shut down yet
-	addonPhase := func() error {
-		_, err := k.c.RunCmd(exec.Command("/bin/bash", "-c", fmt.Sprintf("%s phase addon all --config %s", baseCmd, conf)))
-		return err
+		klog.Infof("needs reconfigure: apiserver error: %v", err)
+		return true
 	}
-	if err = retry.Expo(addonPhase, 100*time.Microsecond, 30*time.Second); err != nil {
-		klog.Warningf("addon install failed, wil retry: %v", err)
-		return errors.Wrap(err, "addons")
+	if st != state.Running {
+		klog.Infof("needs reconfigure: apiserver in state %s", st)
```

Comment:
Is state.State an int (type State int), can we use %d instead?

Also, do we expect to know what the state ID means by the number itself? Or should we lookup a more meaningful error, like state ID 1 = "missing xyz", and display this to the user?

Reply:
Changed to `st.String()`